test/mypy_visitor.py

```diff
@@ -0,0 +1,191 @@
+            finder = self.orig_hook(path)
+            if isinstance(finder, FileFinder):
+                # Move pure python file loaders to the front
+                finder._loaders.sort(  # type: ignore
+                    key=lambda pair: 0 if pair[0] in (".py", ".pyc") else 1
+                )
+
+            return finder
+
+    sys.path_hooks = [PreferPureLoaderHook(h) for h in sys.path_hooks]
+    sys.path_importer_cache.clear()
+
+    yield
+
+    # Restore the previous behaviour
+    original_hooks = []
+    for hook in sys.path_hooks:
+        assert isinstance(hook, PreferPureLoaderHook)
+        original_hooks.append(hook.orig_hook)
+    sys.path_hooks = original_hooks
+    sys.path_importer_cache.clear()
+
+
+@contextmanager
+def pure_python_mypy() -> Iterator[None]:
+    """
+    Inside this context, all mypy related imports are done with the pure python
+    versions.
+
+    Any existing mypy module that was imported before needs to be reimported
+    before use within the context.
+
+    Upon exiting, the previous implementations are restored.
+    """
+
+    def loaded_mypy_modules() -> Iterator[str]:
+        """Covenient block to get names of imported mypy modules"""
+        for mod_name in sys.modules:
+            if mod_name == "mypy" or mod_name.startswith("mypy."):
+                yield mod_name
+
+    # First, backup all imported mypy modules and remove them from sys.modules,
+    # so they will not be found in resolution
+    saved_mypy = dict()
+    for mod_name in list(loaded_mypy_modules()):
+        saved_mypy[mod_name] = sys.modules[mod_name]
+        del sys.modules[mod_name]
+
+    with prefer_pure_python_imports():
+        # After the modules are clean, ensure the newly imported mypy modules
+        # are their pure python versions.
+        # - Pure python: methods are FunctionType
+        # - Native: methods are MethodDescriptorType
+        from mypy.traverser import TraverserVisitor
+
+        assert isinstance(
+            typing.cast(FunctionType, TraverserVisitor.visit_var), FunctionType
+        )
+
+        # Give back control
+        yield
+
+    # We're back and this is where we do cleanup. We'll remove all imported
+    # mypy modules (pure python) and restore the previously backed-up ones
+    # (allegedly native implementations)
+    for mod_name in list(loaded_mypy_modules()):
+        del sys.modules[mod_name]
+
+    for mod_name, module in saved_mypy.items():
+        sys.modules[mod_name] = module
+
+
+def _get_class_globals(clazz: type, localns: Namespace) -> Namespace:
+    """
+    Get the globals namespace for the full class hierarchy that starts in
+    clazz.
+
+    This follows the recommendation of PEP-563 to resolve stringified type
+    annotations at runtime.
+    """
+    all_globals = localns.copy()
+    all_globals.update(localns)
+    for base in inspect.getmro(clazz):
+        all_globals.update(vars(sys.modules[base.__module__]))
+    return all_globals
```

Comment:
Few things:

* I would use `_class` or `class_` instead of `clazz` (more readable IMO)
* The `all_globals.update(localns)` line shouldn't be needed, since we already are `copy()`ing on the line above

Reply:
Thanks, I hated clazz, and this actually made me think of a more specific name instead.

WRT the `update`: 🤦 . Yep.